File: testdata/fixtures/ruby-dirty/domain/order.rb
# Domain layer: should stay pure (std only) — but here it reaches UP into the
# service layer, a layering inversion the rules forbid.
require "securerandom"

require_relative "../service/notify"

module Domain
  class Order
    attr_reader :id, :total

    def initialize(total)
      @id = SecureRandom.uuid
      @total = total
    end
  end

  def self.new_order(total)
    order = Order.new(total)
    Service.notify(order.id)
    order
  end
end

File: testdata/fixtures/python-dirty/domain/order.py
# Domain layer: should stay pure (std only) — but here it reaches UP into the
# service layer, a layering inversion the rules forbid.
import uuid
from dataclasses import dataclass

from service.notify import notify


@dataclass
class Order:
    id: str
    total: int


def new_order(total: int) -> Order:
    order = Order(id=str(uuid.uuid4()), total=total)
    notify(order.id)
    return order
